namesys/namesys.go

```diff
@@ -72,38 +95,100 @@ func (ns *mpns) resolveOnce(ctx context.Context, name string) (path.Path, error)
 		return "", ErrResolveFailed
 	}
 
-	for protocol, resolver := range ns.resolvers {
-		log.Debugf("Attempting to resolve %s with %s", segments[2], protocol)
-		p, err := resolver.resolveOnce(ctx, segments[2])
-		if err == nil {
-			if len(segments) > 3 {
-				return path.FromSegments("", strings.TrimRight(p.String(), "/"), segments[3])
-			} else {
-				return p, err
+	makePath := func(p path.Path) (path.Path, error) {
+		if len(segments) > 3 {
+			return path.FromSegments("", strings.TrimRight(p.String(), "/"), segments[3])
+		} else {
+			return p, nil
+		}
+	}
+
+	// Resolver selection:
+	// 1. if it is a multihash resolve through "pubsub" (if available),
+	//    with fallback to "dht"
+	// 2. if it is a domain name, resolve through "dns"
+	// 3. otherwise resolve through the "proquint" resolver
+	key := segments[2]
+
+	_, err := mh.FromB58String(key)
+	if err == nil {
+		res, ok := ns.resolvers["pubsub"]
```

Comment:
Another approach is to resolve with both resolvers in parallel and use the IPNS record with the largest sequence number.